cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -134,6 +146,56 @@
     this.tasks = new LinkedList<>();
     this.metricsCollectionService = metricsCollectionService;
     this.programCompletionNotifiers = Collections.emptySet();
+    this.cConf = cConf;
+    this.txRunner = transactionRunner;
+    this.store = store;
+  }
+
+  @Override
+  protected void doStartUp() throws Exception {
+    super.doStartUp();
+    AtomicBoolean completed = new AtomicBoolean();
+    AtomicReference<AppMetadataStore.Cursor> cursorRef = new AtomicReference<>(AppMetadataStore.Cursor.EMPTY);
+    RetryStrategy retryStrategy = RetryStrategies.fromConfiguration(cConf, "system.runtime.monitor.");
+    int limit = cConf.getInt(Constants.RuntimeMonitor.INIT_BATCH_SIZE);
+    long startTs = System.currentTimeMillis();
+
+    try {
+      while (!completed.get()) {
+        Retries.runWithRetries(() ->TransactionRunners.run(txRunner, context -> {
+          AppMetadataStore store = AppMetadataStore.create(context);
+          completed.set(true);
+          store.scanActiveRuns(cursorRef.get(), limit, (cursor, runRecordDetail) -> {
+            completed.set(false);
```

Comment:
Does it mean it will never finish the while if there are active runs after startTs?

Reply:
Resolved based on offline discussion by adding scanActiveRuns to Store class.